### PR TITLE
Add a parser for the Flutter Markup Language

### PR DESCRIPTION
**Pull request by abarth**
This patch just contains the syntax. I'm still working on the semantics.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @eseidelGoogle

**Comment by tain335** [comment]
Is this meant to use a markup language to make a widget?

**Comment by abarth** [comment]
> Is this meant to use a markup language to make a widget?

Yeah, it's a bit of an experiment.  I haven't worked out all the details yet.

File: packages/fml/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+name: fml
```
Comment (Hixie) [inline packages/fml/pubspec.yaml]: we should probably pick another name...

**Comment by Hixie** [comment]
I don't think I'd use angle brackets (they're hard to type on many keyboard layouts). I don't think I'd use untyped attributes. I've no idea what you're going to do with this or what problem it solves. It needs tests. But LGTM, I'm curious to see where it goes. :-D

File: packages/fml/pubspec.yaml (hunk 1 above)
Comment (abarth) [inline packages/fml/pubspec.yaml]: Aw, the name is why I got interested in this again.  :)

Comment (Hixie) [inline packages/fml/pubspec.yaml]: it's wtf all over again. :-P

**Comment by abarth** [comment]
> I don't think I'd use angle brackets (they're hard to type on many keyboard layouts).

Yeah, maybe I should experiment with the syntax a bit more.

> I don't think I'd use untyped attributes.  I've no idea what you're going to do with this

Maybe I should have lead with a design proposal.  The idea is to use the markup at compile time to generate Dart code.  The markup gets parsed and sent to the ParserSink, which dispatches further according to the tagName of the element.

The code that it dispatches to will be autogenerated from the source tree, but a widget will be able to customize how it parses.  For example, Text will have a custom parser to use its interior text node as its string.  The example below also shows custom parsing for Color and FractionalOffset.  You can insert Dart code directly using mustaches (i.e., `{{` and `}}`), but that has to be the entirety of an attribute (or text node).

I'm least sure about how to handle widgets that take multiple named child widgets.  My current idea is to use the text nodes that precede the elements as labels.  The autogenerator would recognize `child` as a common pattern and let you omit the label.

There's an example that shows these ideas working together:

```
<Scaffold>
  toolbar: <Toolbar>
    center: <Text>Example app</Text>
  </Toolbar>
  body: <Align alignment="1.0, 0.0">
    <RaisedButton onPressed="{{ _handlePressed }}" color="#F44336">
      <Text>Press me!</Text>
    </RaisedButon>
  </Align>
</Scaffold>
```

> or what problem it solves.

The problem I'm trying to solve is that writing build functions is frustrating because you need to get all the parens and commas and such just right.

> But LGTM, I'm curious to see where it goes.

Maybe this work is too speculative for master.  I could work on it in another branch if you'd prefer.

**Comment by Hixie** [comment]
I'm very happy for this to be on master, sorry if that came over as negative. It was supposed to come over as gung ho despite my concerns. :-)

> The problem I'm trying to solve is that writing build functions is frustrating because you need to get all the parens and commas and such just right.

Aren't you going to have to get the angle brackets and quotes just right too?

> Yeah, maybe I should experiment with the syntax a bit more.

FWIW, I thought the syntax you had described a few months ago was promising.

**Comment by Hixie** [comment]
BTW, I've been thinking of working on a markup language myself, but the problem space I was planning on attacking was styling of paragraphs, inlining images, that kind of thing. Basically, document markup.

**Comment by abarth** [comment]
> It was supposed to come over as gung ho despite my concerns. :-)

Ah, ok.  :)

> Aren't you going to have to get the angle brackets and quotes just right too?

That seems to be easier for folks.  I don't have a complete theory as to why.  There's also some amount of error-correction built into the parser, similar to HTML.  For example, if you screw up and leave off the `</Align>`, the whole thing will be fine.

> FWIW, I thought the syntax you had described a few months ago was promising.

This example would become:

```
Scaffold:
  toolbar: Toolbar:
    center: Text: "Example app"
  body: Align:
    alignment: (1.0, 0.0)
    child: RaisedButton:
      onPressed: {{ _handlePressed }}
      color: #F44336
      child: Text: "Press me!"
```

It feels like slightly too make `:`s, but there's much less other punctuation.

> BTW, I've been thinking of working on a markup language myself, but the problem space I was planning on attacking was styling of paragraphs, inlining images, that kind of thing. Basically, document markup.

As long as you liked the tokenization, you can plug that sort of thing into this system.  Basically, StyledText would take over the processing of tokens within itself, so you could have `<b>` and `<i>` elements that it would recognize internally and translate into TextStyles instead of widgets.

**Comment by geevi** [comment]
Have you guys looked at [QML](https://en.wikipedia.org/wiki/QML). Its more general with support for bindings and animation. Also there is angular2's syntax for binding and callbacks.

**Comment by abarth** [comment]
> Have you guys looked at QML

Yeah, the semantics aren't right for our application but the syntax has some good ideas.  For Flutter, it might look something like

```
Scaffold {
  toolbar: Toolbar {
    center: Text {
      "Example app"
    }
  }
  body: Align {
    alignment: (1.0, 0.0)
    child: RaisedButton {
      onPressed: _handlePressed
      color: #F44336
      child: Text {
        "Press me!"
      }
    }
  }
}
```

I guess it puts you back in the business of needing to balance parenthesis, just now they're braces.  Here's the existing pure-Dart syntax for reference:

```
new Scaffold(
  toolbar: new Toolbar(
    center: new Text(
      "Example app"
    )
  ),
  body: new Align(
    alignment: new FractionalOffset(1.0, 0.0),
    child: new RaisedButton(
      onPressed: _handlePressed,
      color: new Color(0xFFF44336),
      child: new Text(
        "Press me!"
      )
    )
  )
);
```

Mostly we've just exchanged `(` for `{` and removed the `new` and `,` punctuation.

**Comment by devoncarew** [comment]
Somewhat related, I wrote a small tool to parse the Flutter widget hierarchy and write out the class and property info as json:

https://gist.github.com/devoncarew/a7bcc7389688c9e6e6ef

It's all the widgets, their names and docs, and the fields defined on them.

**Comment by a14n** [comment]
And if [`new` is made optional](dart-lang/sdk#20750) the pure-Dart syntax would be even better:

```dart
Scaffold(
  toolbar: Toolbar(
    center: Text(
      "Example app"
    )
  ),
  body: Align(
    alignment: FractionalOffset(1.0, 0.0),
    child: RaisedButton(
      onPressed: _handlePressed,
      color: Color(0xFFF44336),
      child: Text(
        "Press me!"
      )
    )
  )
);
```

**Comment by sethladd** [comment]
Man, I wish we could just change Dart's syntax so we don't have to invent a new language, analyzer support, tools, syntax highlighting in editors, etc, etc. We're going to want code completion, warnings when the bindings syntax doesn't match the Dart method name, rename refactoring across Dart code and FML, etc.

So, my question is, what _if_ Dart syntax was:

```
Scaffold(
  toolbar: Toolbar(
    center: Text(
      "Example app"
    )
  ),
```

That personally looks really clean to me. Curious what others think.

**Comment by abarth** [comment]
> That personally looks really clean to me. Curious what others think.

IMHO, there would still be too many commas and parens that need to be balanced.

**Comment by eukreign** [comment]
I think it might make more sense to add a serialization format targeted at WYSIWYG tools and to keep Dart as the main method of defining widgets manually/programmatically.

The Dart syntax and the proposed declarative syntax are too similar to justify the effort but having a nice alternative serialization format with an accompanying programmatic API for others to build UI editors, automation/build scripts and other tools would be pretty cool.

**Comment by abarth** [comment]
I'm going to close this pull request.  We should probably start out with a design-doc for how we want this to work and integrate with WYSIWYG tools.

**Comment by krisgiesing** [comment]
This is very close to what I was recommending in the monocle design doc (go/tq-monocle).  We're working on a format right now that's syntactically similar to JSON and actually looks quite a bit like your example in comment 11 (see go/tq-adin).

**Pull request closed without merging**